src/serac/numerics/functional/integral.hpp

```diff
@@ -0,0 +1,217 @@
+#pragma once
+
+#include <array>
+#include <memory>
+
+#include "mfem.hpp"
+
+#include "serac/infrastructure/accelerator.hpp"
+#include "serac/numerics/functional/geometric_factors.hpp"
+#include "serac/numerics/functional/function_signature.hpp"
+#include "serac/numerics/functional/domain_integral_kernels.hpp"
+#include "serac/numerics/functional/boundary_integral_kernels.hpp"
+
+namespace serac {
+
+struct Integral {
+  enum Type
+  {
+    Domain,
+    Boundary,
+    // DG, unimplemented
+    _size
+  };
+
+  static constexpr Type Types[2] = {Domain, Boundary};
+
+  static constexpr std::size_t num_types = Type::_size;
+
+  Integral(Type t, std::vector<uint32_t> trial_space_indices) : type(t), active_trial_spaces(trial_space_indices)
+  {
+    std::size_t num_trial_spaces = trial_space_indices.size();
+    evaluation_with_AD_.resize(num_trial_spaces);
+    jvp_.resize(num_trial_spaces);
+    element_gradient_.resize(num_trial_spaces);
+
+    for (size_t i = 0; i < num_trial_spaces; i++) {
```

Comment:
Some documentation on the `functional_to_integral_` container would greatly improve readability.

Reply:
maybe `functional_to_indegral_index_`?